[2단계 - 자동차 경주] 밍고(성민제) 미션 제출합니다.

MinSungJe opened the pull request:
도밥 안녕하세요!! 밍고입니다.😊
자동차 경주 2단계 미션 제출합니다.
1단계 리뷰 때 너무 상세히 잘 봐주셔서 감사한 마음뿐입니다.🥲

PR글을 적기 앞서 1단계 리뷰 중 [**PR에 담기면 좋을 내용**](https://github.com/woowacourse/javascript-racingcar/pull/334#pullrequestreview-2614677702)을 자세히 적어주셔서 해당 내용을 바탕으로 PR글을 적어보겠습니다.
혹시 PR을 읽으시면서 부족한 점이 보인다면 해당 사항에 대해서도 아낌없는 리뷰 부탁드립니다.
2단계도 잘 부탁드립니다!

---

# 🚀 2단계 - 자동차 경주 리팩터링

## 🎯 리팩터링 요구 사항
- 도메인 로직은 domain/ 하위로, UI 관련 로직은 view/ 하위에서 관리한다.
- domain/ 하위의 모듈은 view/ 하위의 모듈을 의존하지 않아야 한다.
- 도메인 로직 내에서 테스트하기 어려운 부분을 분리하고, 테스트 가능한 부분에 대해서만 단위 테스트를 작성한다.
  - 테스트 코드에서 jest.fn()을 사용하지 않는다. (마찬가지로, mock/spy 등도 이번 미션에서는 사용하지 않는다.)

## 🔧 코드 설계 의도
### Step 1 내 반영하지 못한 리뷰 반영
번호 | 리뷰 내용 | 관련 코멘트 | 반영 커밋
-- | -- | -- | --
1|도메인 객체에서 출력 기능 분리|[#334 (review)](https://github.com/woowacourse/javascript-racingcar/pull/334#discussion_r1956516990)|https://github.com/woowacourse/javascript-racingcar/commit/fcd7cec9a720b899c4a9007c2fd4001f5087596c
2|테스트 명세 명확화|[#334 (review)](https://github.com/woowacourse/javascript-racingcar/pull/334#discussion_r1956520979)|https://github.com/woowacourse/javascript-racingcar/commit/c96be5f2fd127e994c9004a08770ea41c6b872c9

특히 **테스트 명세 명확화**는 step 1에서 여러 번 짚어주신 리뷰 내용이라 더욱 신경써서 작성해보았습니다.

### 도메인과 UI의 구분
❗이번 2단계 미션을 수행하며 가장 신경썼던 부분입니다!
기존에 있던 코드를 다음과 같이 분류하여 리팩토링을 진행하였습니다.
<image src='https://github.com/user-attachments/assets/3a26f6b4-7cac-4b56-a8d7-4a108e64e8f3' width=70%/>

Domain에선 **필드의 정보만 전달**하도록 리팩토링하여 UI와의 의존성을 제거했습니다.
UI에선 Domain 내부의 public 함수를 통해 경주 정보에 접근하고 이를 출력합니다.
실제로 Domain의 정보를 가지고 UI를 통해 사용자에게 출력하는 로직은 App.js 내부에서 구현되었습니다.

각 기능에 대해 리팩토링한 커밋은 아래와 같습니다.
- https://github.com/woowacourse/javascript-racingcar/commit/fcd7cec9a720b899c4a9007c2fd4001f5087596c Car 객체에서 출력 양식 반환 함수 제거 / Car 정보 반환 함수 구현 
- https://github.com/woowacourse/javascript-racingcar/commit/8405631ce7b93ba68fb333e3cb604316724f6cc5 Racing 객체 도메인 분류, Racing에서 경기 현황 및 우승자를 출력하는 함수를 다른 위치로 분리
- https://github.com/woowacourse/javascript-racingcar/commit/a3c30d9040677f7da71bc877c0e8ff22936e67d4 입력을 담당하는 로직을 App.js에서 InputView로 분리, UI로 분류
- https://github.com/woowacourse/javascript-racingcar/commit/49d7e9f80e8ffbc823658a89f0530ccabeb21b7f Domain과 UI를 통합하는 로직을 App.js에서 구현

Domain과 UI의 분류를 많이 고민했는데 해당 분류가 잘 되었는지 확인해주시면 감사하겠습니다.🙂

MinSungJe committed:
docs: 리팩토링 요구 사항 README 반영

MinSungJe committed:
refactor: Car 내 필드 공개 범위 수정

필드의 공개 범위를 수정하고 필드 정보를 반환하는 함수 구현
이를 통해 외부 사용 및 테스트 코드 작성 가능

MinSungJe committed:
refactor: Car 객체 domain 분류

MinSungJe committed:
refactor: Racing 객체 도메인 분류

Racing 속 UI로 판단되는 함수(start())를 printMessage.js로 이동(이후 refactor 필요)
우승자를 출력할 때 바로 출력형식으로 바꾸는 게 아니라 배열로 반환하도록 변경
불필요한 count 필드 삭제

이를 통해 Racing 객체가 도메인 역할만 하도록 기능을 분류함

MinSungJe committed:
refactor: 기존 UI 관련 파일 위치 변경

MinSungJe committed:
refactor: InputView 분리

사용자의 입력을 받는 App.js 속 로직을 InputView로 분리

MinSungJe committed:
feat: Domain과 UI 통합 로직 App에 구현

MinSungJe committed:
chore: printMessage.js -> OutputView.js로 이름 변경

MinSungJe committed:
fix: 최종 우승자 출력 시 여백 추가

MinSungJe committed:
test: InputValidator 테스트 import 문 오류 수정

MinSungJe committed:
test: 테스트 명세를 명확하게 변경

Creative-Lee commented on src/view/InputView.js:
```diff
@@ -0,0 +1,23 @@
+    InputValidator.carNames(names);
+
+    return names.map((name) => new Car(name, 0));
+  },
```
InputView (view로직) Car class(domain로직)를 직접 생성하고 반환하고 있네요.
view와 도메인이 분리되지 않았다고 할 수 있습니다. 

InputView는 사용자의 입력을 반환하는 역할만 하고,
Car 인스턴스 생성은 또다른 도메인 Class(Racing)가 담당하도록 개선해 봅시다.

MinSungJe replied:
https://github.com/woowacourse/javascript-racingcar/pull/389/commits/2eb5caf7b022bbcdcd8b4793f77e58a69042541b
Racing을 만들 때 CarList가 아닌 CarNames를 받고, Racing의 constructor에서 Car 객체를 만들도록 변경하여 반영 완료했습니다!

Creative-Lee commented on src/view/InputView.js:
```diff
@@ -0,0 +1,23 @@
+  async enterCarNames() {
+    const inputName = await readLineAsync(GAME_MESSAGE.ENTER_CAR_NAMES);
+    const names = inputName.split(SEPARATOR);
+    InputValidator.carNames(names);
```
2가지 개선 포인트가 있으니 하나씩 개선해 봅시다.

1. 사용자의 입력을 검증하는 로직또한 도메인 로직이 아닐까요? ('자동차의 이름은 5글자 이하여야 한다'는 게임과 밀접한 '정책'이죠?)
사용자의 입력을 최초 검증하는 로직은 사용자의 입력을 필요로 하는 곳에서 수행하도록 개선해봅시다.

1번 개선이 끝난 후 이어서 아래 코멘트로 넘어와주세요!

---

2. 입력받는 부분을 `클라이언트`, 도메인 객체(Car, Racing)를 `서버`라고 생각해보세요.
서버가 클라이언트만 믿고 아무런 검증없이 서버로직을 수행해도 될까요?
사용자 입력에 대해 최초에 검증을 하지만, 만약 다른 개발자가 input에 대한 검증을 깜빡하고 new Car("여섯글자이름") 으로 자동차를 만들려 시도한다면 어떻게 될까요? -> 아무런 에러없이 정책을 어기는 객체가 생성될겁니다.
Car클래스 자체에서도 검증이 이루어진다면 훨씬 더 안전하지 않을까요?
그렇게 된다면, 어떤 개발자가 아무 생각없이 사용자 입력을 받자마자 인스턴스를 초기화 한다고 하더라도 비즈니스 정책(자동차의 이름은 n글자 이하)이 지켜질 것 같아요.

MinSungJe replied:
https://github.com/woowacourse/javascript-racingcar/pull/389/commits/85431614caa853c5ca54c8b5a34a5c0e66676acb
1번 사항은 InputValidator를 domain으로 옮기고 domain과 UI를 같이 이용하는 함수를 App.js에서 새로 만들어 반영했습니다.

https://github.com/woowacourse/javascript-racingcar/pull/389/commits/78c0e675c7d2d4162a04348cdd166d5e4f078cb6
2번 사항 관련해서는 지금까지 필요없는 이중검증을 하는 거라 생각해 하지 않았는데, 이렇게 타당한 이유를 들어 설명하시니 필요성에 대해 특히 와닿게 되네요🙂🙂 반영 완료했습니다!

Creative-Lee commented on src/view/OutputView.js:
```diff
@@ -0,0 +1,30 @@
+    console.log();
+  },
+
+  printRaceStatus(racing) {
+    racing.carList.forEach((car) => {
+      const { name, position } = car.getCarStatus();
+      OutputView.printMessage(`${name} : ${FORWARD_DASH.repeat(position)}`);
+    });
+  },
```
outputView (view로직)이 racing 인스턴스를 인자로 전달받아 도메인객체의 메서드(getCarStatus)를 직접 호출하고 있습니다. 이 역시 view와 도메인이 분리되지 않았다고 할 수 있습니다.

outputView는 **문자열 및 숫자만 넘겨받아 이를 출력하는 역할**만 담당하고,

> 실제로 Domain의 정보를 가지고 UI를 통해 사용자에게 출력하는 로직은 App.js 내부에서 구현되었습니다.

밍고가 pr 본문에 적어주신 것처럼 한 스탭만 더 분리해서 
도메인 정보를 가지고 오는 `car.getCarStatus()` 메서드의 호출만 적절한 위치로 변경하면 완벽하게 분리될 것 같아요

MinSungJe replied:
https://github.com/woowacourse/javascript-racingcar/pull/389/commits/79ded6be9fe02b5443988590761faf828e5c271c
App.js에서 필요한 정보를 도메인에서 반환받아 그 정보만 UI에서 출력하도록 변경하여 반영 완료했습니다!

Creative-Lee reviewed (changes requested):
안녕하세요 밍고~ 👋 저 또한 감사합니다 😄 

밍고가 작성한 계층구조 덕분에 App의 전체 구조를 한눈에 파악할 수 있어서 리뷰가 상당히 수월했습니다ㅎㅎ
또 반영한 리뷰들과 기능별 리팩토링 커밋을 잘 첨부해주셔서 변화된 코드를 비교하며 보기도 좋았네요! 👍 (테스트 명세도 상당히 깔끔해졌어요 😄  )  

>Domain과 UI의 분류를 많이 고민했는데 해당 분류가 잘 되었는지 확인해주시면 감사하겠습니다.

코드에서 view와 domain 로직을 잘 분리하고자 하는 노력이 보였습니다 👍 
분리 측면에서 step1 마무리 당시 코드보다 훨씬 잘 나뉘어진 모습이네요!
다만, 일부 분리되지 않은 코드가 보여서 코멘트 남겨드렸으니 확인 부탁드릴게요 😄

MinSungJe committed:
refactor: Car 인스턴스를 Racing에서 생성

Racing에서 CarName 배열을 받아 인스턴스를 생성하도록 변경
기존에 InputView(UI)에서 생성하던 것을 옮겨 도메인/UI 분리

MinSungJe committed:
refactor: OutputView가 racing을 직접 전달받지 않도록 수정

OutputView(UI)가 Racing(Domain)을 직접 전달 받지 않도록 수정
도메인의 특정 값을 App.js에서 호출하여 UI에 전달하도록 수정

MinSungJe committed:
refactor: InputValidator 도메인으로 분류 변경

도메인과 UI를 동시에 사용하는 함수를 App.js에 새로 구현 후 사용

MinSungJe committed:
feat: 도메인 객체를 만들 때 검증 추가

Creative-Lee commented on src/App.js:
```diff
@@ -1,32 +1,46 @@
-    const count = await readLineAsync(GAME_MESSAGE.ENTER_COUNT);
-    InputValidator.count(count);
+  #race(racing, count) {
+    OutputView.printMessage(GAME_MESSAGE.RACING_RESULT);
+
+    for (let i = 0; i < count; i++) {
+      racing.raceOnce();
+      racing.carList.forEach((car) => {
+        const { name, position } = car.getCarStatus();
+        OutputView.printRaceStatus(name, position);
+      });
+      OutputView.printBlank();
+    }
 
-    return count;
+    const winner = racing.getWinner();
+    OutputView.printRaceWinner(winner);
   }
 }
```
`사용자가 입력한 라운드 수 입력만큼 raceOnce() 해야한다.` 는 도메인 로직이라고 생각하는데요~~!
이 로직은 밍고의 코드에선 Racing.js 가 관리해야하는 부분이 아닐까요?
밍고는 어떻게 생각하시나요~?

MinSungJe replied:
저는 한번 경주를 할 때마다(Domain) 경주의 현황을 출력(UI)하는 요구사항을 만족시켜야 하니까 `App.js`에서 두 로직을 불러오는 것이 자연스럽다고 생각했어요! 경주를 몇 번 실행하는 것만 있으면 Domain에서 처리해도 괜찮지만 경주마다 현황을 출력해야 하니까요.

Creative-Lee replied:
밍고의 의견을 정리하면, 
domain로직임에는 동의하나, ui출력 요구사항을 만족하기위해 app에 두었다 인것같아요

음... 'ui출력 요구사항을 만족하기 위해 domain로직을 밍고가 정의한 controller 에 두엇다'는 합당한 사유가 아니라고 생각되어요.

ui 요구사항을 만족하면서도 domain로직이 있어야 할 곳에 있는게 맞지않을까요?
예를들면 매 라운드 결과를 저장해서 한번에 외부에서 출력한다던지, 혹은 (도메인으로 분리된) race 함수가 출력 함수를 onRoundEnd 등의 인자로 받아 실행하도록 할수도 있을거같고요.
@MinSungJe

MinSungJe replied:
매 라운드 결과를 저장해서 한번에 외부에서 출력하는 방식으로 구현해보겠습니다! 마지막으로 고민이 되는건 도메인 내 함수의 이름인데, 경주를 하는 함수니 `racing.start()`로 이름을 지어야 할지 아니면 출력할 상태를 얻는 함수니 `racing.getStatus()`로 이름을 지어야 할 지 고민입니다.

저는 이 함수를 통해 출력할 상태를 반환한다는 걸 이름에 반영하고 싶어 `racing.getStatus()`로 이름을 짓고 싶은데, 도밥은 어떻게 생각하는 지 궁금합니다😢

```js
경주를하는로직이자경기결과를얻는함수(count) {
    const raceStatus = [];

    for (let i = 0; i < count; i++) {
      const turn = [];
      this.raceOnce();
      this.carList.forEach((car) => {
        const { name, position } = car.getCarStatus();
        turn.push({ name, position });
      });
      raceStatus.push(turn);
    }

    return raceStatus;
  }
```

MinSungJe replied:
https://github.com/woowacourse/javascript-racingcar/pull/389/commits/82baac69cde9f8a8d6820e0b11e9b8b8e05fe6d5
Racing에서 모든 라운드 결과를 저장 후 반환하도록 하고, 이를 View에 전달하여 결과를 출력하도록 리팩토링하여 반영했습니다!

리팩토링 하다보니 함수의 네이밍 관련해서 의문이 있었는데요. 도메인에서 새로 리팩토링한 함수의 기능은 `경주를 진행하며 매 라운드 결과를 저장해 이를 반환`하는 역할입니다. 저는 **매 라운드 결과를 반환하는** 기능에 집중해 이 함수를 `racing.getStatus()`로 이름지었습니다. 하지만 **경주를 실제로 하는** 기능을 이름에 담고 싶다면 `racing.start()` 등과 같이 이름을 지을 수도 있어 보입니다. 어떤 이름이 코드를 읽는 입장에서 더 적절할지 도밥의 의견이 궁금합니다.😢
```js
// App.js
  #race(racing, count) {
    OutputView.printMessage(GAME_MESSAGE.RACING_RESULT);

    // 깔끔하지만 경주가 실제 어디서 진행되고 있는 지 알기 어려울 수 있어 보임.
    const raceStatus = racing.getStatus(count);
    OutputView.printRaceStatus(raceStatus);

    const winner = racing.getWinner();
    OutputView.printRaceWinner(winner);
  }
```

Creative-Lee replied:
항상 네이밍이 말썽이죠 😢 가볍게 코멘트 달아보면, 
`racing.getStatus` 라는 메서드 명만 보고 경주가 시작된다고 예상하기 어려운것 같기는 합니다. 
밍고의 생각처럼 start()나 race() 같은 네이밍이면 어떨까 싶습니다. 
start()나 race() 안에서 raceOnce()를 실행하는것도 어색하지는 않고요.
이후에 메서드에서 반환된 결과값은 `roundResults` 같은것도 좋겠네요

Creative-Lee reviewed (changes requested):
반영된 사항 확인했습니다 ㅎㅎ 커밋단위도 잘 나뉘어 있어서 좋네요~!
마지막 코멘트 하나 남겨봤습니다! 이것만 고민해보시고 요청주세요~!(밍고의 의도가 맞다면 수정없이 요청주시면 되겠습니다~!) 
끝이 보이네요!  화이팅!

MinSungJe committed:
refactor: 실제 경주를 하는 기능을 도메인으로 분류

도메인인 Racing에서 모든 경주 로직이 진행되도록 변경
Racing의 getStatus로 경주결과를 얻고 이를 view에 전달해 출력하도록 설계

MinSungJe committed:
refactor: 하나의 라운드를 출력하는 함수 분리

이중 반복문을 지양하기 위해 printRaceStatus 내 하나의 라운드를 출력하는 함수 분리

Creative-Lee reviewed (approved):
최최최종.hwp 리뷰 반영까지 확인했습니다 ㅎㅎ
여러 부분에서 코드가 충분히 개선되었다고 생각해요~!
PR에 담으면 좋은 부분들도 바로바로 step2에서 반영해주셔서 감사했습니다.

우테코 첫 미션 정말 고생많으셨네요~ 👍  여기서 이번 미션은 마무리하도록 하겠습니다.
다음에 또 봐요 밍고 👋
